src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1078,246 @@ begin
 end
 
 end matrix_adjoints
+
+namespace bilin_form
+
+section orthogonal
+
+/-- The orthogonal complement of a submodule `N` with respect to some bilinear form is the set
+  of elements which are orthogonal to all elements of `N`. -/
+def orthogonal (B : bilin_form R M) (N : submodule R M) : submodule R M :=
+{ carrier := { m | ∀ n ∈ N, is_ortho B n m },
+  zero_mem' := λ x _, is_ortho_zero_right x,
+  add_mem' := λ x y hx hy n hn,
+    by rw [is_ortho, add_right, show B n x = 0, by exact hx n hn,
+        show B n y = 0, by exact hy n hn, zero_add],
+  smul_mem' := λ c x hx n hn,
+    by rw [is_ortho, smul_right, show B n x = 0, by exact hx n hn, mul_zero] }
+
+variables {N L : submodule R M}
+
+@[simp] lemma mem_orthogonal_iff {N : submodule R M} {m : M} :
+  m ∈ B.orthogonal N ↔ ∀ n ∈ N, is_ortho B n m := iff.rfl
+
+lemma orthogonal_le (h : N ≤ L) : B.orthogonal L ≤ B.orthogonal N :=
+λ _ hn l hl, hn l (h hl)
+
+lemma le_orthogonal_orthogonal (hB : refl_bilin_form.is_refl B) :
+  N ≤ B.orthogonal (B.orthogonal N) :=
+λ n hn m hm, hB _ _ (hm n hn)
+
+end orthogonal
+
+/-- The restriction of a bilinear form on a submodule. -/
+@[simps apply]
+def restrict (B : bilin_form R M) (W : submodule R M) : bilin_form R W :=
+{ bilin := λ a b, B a b,
+  bilin_add_left := λ _ _ _, add_left _ _ _,
+  bilin_smul_left := λ _ _ _, smul_left _ _ _,
+  bilin_add_right := λ _ _ _, add_right _ _ _,
+  bilin_smul_right := λ _ _ _, smul_right _ _ _}
+
+/-- The restriction of a symmetric bilinear form on a submodule is also symmetric. -/
+lemma restrict_sym (B : bilin_form R M) (hB : sym_bilin_form.is_sym B)
+  (W : submodule R M) : sym_bilin_form.is_sym $ B.restrict W :=
+λ x y, hB x y
+
+/-- A nondegenerate bilinear form is a bilinear form such that the only element that is orthogonal
+  to every other element is `0`. -/
+def nondegenerate (B : bilin_form R M) : Prop :=
+∀ m : M, (∀ n : M, B m n = 0) → m = 0
```

Comment:
I thought about it a little and I'm not sure this is the right definition for `nondegenerate`.  It's asymmetric:  there is an alternative condition
`∀ m : M, (∀ n : M, B n m = 0) → m = 0`
(In finite dimension either of these conditions implies the other.)

I found at least one source, section 8.3 of
http://pub.math.leidenuniv.nl/~luijkrmvan/linalg2/2015/LinAlg2-2015.pdf
giving, as the definition of `nondegenerate`, that _both_ these conditions should hold.

Would you like to start a thread on Zulip asking what the correct definition should be?

Reply:
Sure!

Reply:
Discussion here: 
https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/nondegenerate.20bilinear.20form